spec/an_localize_libs_action_spec.rb

```diff
@@ -0,0 +1,54 @@
+require 'spec_helper'
+
+describe Fastlane::Actions::AnLocalizeLibsAction do
+  # This test is more of a way of ensuring `run_described_fastlane_action` handles array
+  # of hashes properly than a comprehensive test for the
+  # `an_localize_libs_action` action.
+  #
+  # Please consider expanding this test if you'll find yourself working on its
+  # action.
+  it 'merges the strings from the given array into the given main strings file' do
+    in_tmp_dir do |tmp_dir|
+      app_strings_path = File.join(tmp_dir, 'app.xml')
+      File.write(app_strings_path, android_xml_with_content('<string name="a_string">test from app</string>'))
+
+      lib1_strings_path = File.join(tmp_dir, 'lib1.xml')
+      File.write(lib1_strings_path, android_xml_with_content('<string name="a_lib1_string">test from lib 1</string>'))
+
+      lib2_strings_path = File.join(tmp_dir, 'lib2.xml')
+      File.write(lib2_strings_path, android_xml_with_content('<string name="a_lib2_string">test from lib 2</string>'))
+
+      run_described_fastlane_action(
+        app_strings_path: app_strings_path,
+        libs_strings_path: [
+          { library: 'lib_1', strings_path: lib1_strings_path, exclusions: [] },
+          { library: 'lib_2', strings_path: lib2_strings_path, exclusions: [] },
+        ]
+      )
+
+      # Notice the extra indentation in the library strings. The action doesn't
+      # modify the app's strings content indentation, but it applies its own
+      # standard to the values read from the given library strings
+      expected = <<~XML
+        <string name="a_string">test from app</string>
+          <string name="a_lib1_string">test from lib 1</string>
+          <string name="a_lib2_string">test from lib 2</string>
+      XML
```

Comment:
I think that's simply due to the indentation in the test using 2 spaces, as it's the default style for this codebase, and the action indenting XML using 4 spaces.

I found that interesting and left it there. If it makes for a confusing test, maybe we could update the examples to all use 4 spaces and leave a note about it being the "default"?

I did a quick search and apparently there is no recommended indentation level for XML. The spec mentions [white space handling](https://www.w3.org/TR/2008/REC-xml-20081126/#sec-white-space):

> In editing XML documents, it is often convenient to use "white space" (spaces, tabs, and blank lines) to set apart the markup for greater readability. Such white space is typically not intended for inclusion in the delivered version of the document. On the other hand, "significant" white space that should be preserved in the delivered version is common, for example in poetry and source code.

I'm guessing because most indentation is only meant for readability, then it's up to each author (or formatter author) to choose what they find appropriate.

Reply:
Ah so this would in practice be an artefact of the tests and how the content used for those XML tests are inlined and assembled in the specs themselves (as opposed to if you were using fixture files) then? While in real use case in client apps there's no such indentation discrepancy between the tags from the app and the ones from the libs?

If that's so then yeah better rewrite the tests to make this less confusing and more importantly make it match reality then 😅

Reply:
Ah wait yeah actually I just realized why that is. This is because the first line of your `content` already gets 2 extra spaces from the heredoc itself! Which is why you have to remove 2 spaces from the first line of your content here to compensate.

So this is just misleading and there is in fact no indentation difference in the XML itself.

You could make it less confusing by adding the 2 spaces on the first line here (to match the rest of the lines), and removing the 2 extra leading spaces in your helper method that includes it in the heresoc block.

Reply:
Mmmh actually re-checked your heredoc and the 2 leasing spaces I thought were there aren't 🤨 so scratch what I just said. Still, pretty strange and worth looking into…

Reply:
I took a bit of digging, but I discovered the difference in indentation to be due to the code adding lines from the XML adding 4 spaces. See:
 https://github.com/wordpress-mobile/release-toolkit/blob/da405f2c5ca90d696857ba2ad01da2753daa60dc/lib/fastlane/plugin/wpmreleasetoolkit/helper/android/android_localize_helper.rb#L67

I opened https://github.com/wordpress-mobile/release-toolkit/pull/332 to hopefully make the test clearer.

Reply:
Yeah I remember that our code hardcodes 4 spaces like that, what I found unclear is not why we'd have 4 spaces indentation (that I expected) but instead why the first line With have a different indentation than the rest of the lines.

Your other PR is clearer, thanks for opening it, and it actually does match exactly what I expected the test (and expected XML) would look like from the start so that change in #332 definitively makes more sense to me.

That being said I'm still puzzled as to how the code in this present PR even passed before in its current state (without #332) then. Because the change you made in #332 consists of removing 2 spaces from all the lines of what `content` looked like **except the first line**, while in the code where you later inject that `content` into the XML heredoc you added 2 spaces (switching from 2 to 4 leading spaces on each line) to **all** the lines. So the 2 extra spaces in `indented_content` balance the 2 spaces removed on `content` just like I expect… except for that first line of `content` which gets 2 extra space from the `map` but didn't get 2 spaces removed from `content` to balance.

If I'm not missing anything from ☝️ (but I'm sure I do otherwise that doesn't make sense) that would mean that the string you use for the expectation changed between this PR and #332 (gaining 2 leading spaces only on the first line, as the rest of the lines balance out) yet your specs were green both before and after that change on the first line of the expectation string?!

I think I need some rest, I must be missing something obvious…